PSReadLine/Position.cs

```diff
@@ -35,5 +34,55 @@ private static int GetBeginningOfLinePos(int current)
 
             return newCurrent;
         }
+
+        /// <summary>
+        /// Returns the position of the end of the logical line
+        /// as specified by the "current" position.
+        /// </summary>
+        /// <param name="current"></param>
+        /// <returns></returns>
+        private static int GetEndOfLogicalLinePos(int current)
+        {
+            var newCurrent = current;
+
+            for (var position = newCurrent; position < _singleton._buffer.Length; position++)
+            {
+                if (_singleton._buffer[position] == '\n')
+                {
+                    break;
+                }
+
+                newCurrent = position;
+            }
+
+            return newCurrent;
+        }
+
+        /// <summary>
+        /// Returns the position of the first non whitespace character in
+        /// the current logical line as specified by the "current" position.
+        /// </summary>
+        /// <param name="current">The position in the current logical line.</param>
+        private static int GetFirstNonBlankOfLogicalLinePos(int current)
+        {
+            var beginningOfLine = GetBeginningOfLinePos(current);
+
+            var newCurrent = beginningOfLine;
+
+            while (IsBlank(newCurrent))
+                newCurrent++;
+
+            return newCurrent;
+        }
+
+        private static bool IsBlank(int newCurrent)
+        {
+            var c = _singleton._buffer[newCurrent];
+
+            return Char.IsWhiteSpace(c)
+                && c != '\r'

```

Comment:
It looks like you would more faithfully emulate `vim` if you used `char.IsControl(c)` instead.

Reply:
I think my code is correct.

If I were to exclude all control characters, I would also exclude TAB which should be considered a "blank" character in that case.

Besides, `Char.IsWhiteSpace` handles many control characters but I just want to exclude new line characters. For completeness, this code excludes `\r` which cannot probably be input via PowerShell, though.

Reply:
I wouldn't say your code is wrong, but I quickly tried a few random control characters in vim and `_` skipped over them. I did think it looked weird that vim skipped over characters that render as non-spaces, but it did.

It's not easy to insert control characters today, but it is possible. At some point somebody will ask for `Ctrl+q` support and then it will be easy to insert control characters.

Reply:
Maybe if the method was named `IsVisibleBlank` or `IsHorizontalWhitespace`, then the intention and code is clear.